### ISSUE TITLE
space as digit group separator should not be parsed as decimal point

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#749]
With a journal containing only the below, the amount is parsed as 1. I think (hope) no country uses space for decimal point, so we should prevent this:
```
2018-01-01
  (a)   USD1 000

;$ hledger -f d.j reg amt:1
;2018/01/01                      (a)                       USD1 000      USD1 000
```

And possibly the same issue, the following commodity directive should complain about a missing decimal point (now required), instead it gives a parse error:
```
commodity 1 000  USD

2018-01-01
  (a)   USD1 000

; hledger: d.j:1:12:
; unexpected space
; expecting digit
```
A correct commodity directive does work as intended:
```
commodity 1 000.00  USD

2018-01-01
  (a)   USD1 000

;$ hledger -f d.j reg amt:1000
;2018/01/01                      (a)                   1 000.00 USD  1 000.00 USD
```

**Comment by ony** [comment]
Space have special meaning and cannot be decimal separator:
```haskell
    (firstSep, groups) <- option (Nothing, []) $ do
        leadingDigits <- some digitChar
        option (Nothing, [leadingDigits]) . try $ do
            firstSep <- oneOf sepChars <|> whitespaceChar
            groups <- some digitChar `sepBy1` char firstSep
            return (Just firstSep, leadingDigits : groups)
```

Isn't `USD1` actually treated as a commodity?

**Comment by simonmichael** [comment]
Contrary to the haddock, `firstSep <- oneOf sepChars <|> whitespaceChar` allows a space (and possibly other things in the [unicode Space category](http://hackage.haskell.org/package/base-4.10.1.0/docs/Data-Char.html#t:GeneralCategory)).

**Comment by ony** [comment]
I see problem is in `fromRawNumber` actually we get `(Just ' ', ["1", "000"], Nothing)` same as before we assume that if there is two groups - last one is decimals regardless of what separator were used.

**Comment by simonmichael** [comment]
I've added three tests related to this. I think the numberp/rawnumberp parsing code is a bit hard to follow and we could maybe make it clearer. 

- https://github.com/simonmichael/hledger/blob/master/tests/journal/amounts-and-commodities.test#L96
- http://hledger.org/journal.html#amounts
- http://hledger.org/journal.html#declaring-commodities
- https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Read/Common.hs#L758

**Comment by awjchen** [comment]
Addressing the first example given in the OP:

In http://hledger.org/journal.html#amounts there is a section that explains the observed behaviour.
> ... However, there is some ambiguous way of representing numbers like $1.000 and $1,000 both may mean either one thousand or one dollar. By default hledger will assume that this is sole delimiter is used only for decimals. On the other hand commodity format declared prior to that line will help to resolve that ambiguity differently: ...

This seems to be implemented in the following snippet of `fromRawNumber` (starting on line 716 of `Hledger/Read/Common.hs` on the current master).
```haskell
(mseparator, intparts, mdecimalpoint, frac) =
        case raw of
            -- just a single punctuation between two digits groups, assume it's a decimal point
            (Just s, [firstGroup, lastGroup], Nothing)
                -- if have a decimalHint restrict this assumpion only to a matching separator
                | maybe True (`asdecimalcheck` s) suggestedStyle -> (Nothing, [firstGroup], Just s, lastGroup)

```
Here, `raw` is a triple of `Maybe Char` (the separator character, if any), `[String]` (the digit groups), and `Maybe (Char, String)` (the decmial character and decimal string, if they are detected). `raw` is the output of some inital parsing done in `rawnumberp`.

In the first example, `raw` evaluates to `(Just ' ', ["1", "000"], Nothing)`, matching the pattern of the first case (I have excluded the others), while `suggestedStyle` evaluates to `Nothing`, so the guard evaluates to `True` and we indeed interpret the space as a decimal point.

---

Addressing the second example of the OP:

The parsing fails on the directive on the first line. We are trying to parse `amountp`, which is an alternative of `leftsymbolamountp`, `rightsymbolamountp`, and `nosymbolamount`. The case we want is `rightsymbolamountp`.

`leftsymbolamountp` fails, as expected. But, while trying to parse `rightsymbolamountp`, we fail `rawnumberp` in `Common.hs`. Specifically, it fails on this line:
```haskell
groups <- some digitChar `sepBy1` char firstSep
```
Here, `firstSep` is the space character, and we have just parsed the `1`. Then, this line parses `"000"` for `some digitChar`, then successfully parses one of the two spaces before the commodity symbol `USD`, and is exepcting another `some digitChar` but instead meets a space.

Upon this failure, we backtrack and obtain a nonsensical result of parsing just `1` with no separators, causing failures down the line. We eventually end up in `nosymbolp`, trying and failing to parse `rawnumberp` in the same way, which gives the error we see.

**Comment by simonmichael** [comment]
Re OP example 1, you're right, the observed behaviour is consistent and even documented. I meant to suggest we should change it and break consistency here, knowing that nobody ever uses space as decimal point (?), allowing that 1 to be parsed as part of the commodity symbol instead. I might be wrong though, it might be too confusing. I should link to the context, this [reddit thread](https://www.reddit.com/r/plaintextaccounting/comments/8godxe/numbers_in_commodity_names/) where someone was complaining about having to enclose commodity symbols with numbers in double-quotes.

**Comment by awjchen** [comment]
Oops, I just wasted my breath, haha.

I'm don't have any experience with syntax, but would be possible to introduce a new separator that could be used to

**Comment by awjchen** [comment]
The second example journal file in PR #780 seems like something one would often write if one used whitespace as a decimal separator. If this is the first time we've exposed this parsing bug, is that evidence that no one uses whitespace as a decimal separator?

**Comment by awjchen** [comment]
I take back what I said about disallowing whitespace as decimal points being complex. The manual has a rule that decimal points can only be commas or periods! So perhaps we *should* disallow it, since, in some sense, it's the current behaviour that complicates things by creating an exception to the above rule!

**Comment by simonmichael** [comment]
Good question. Eh.. this gets confusing. I think: 

- nobody uses space as decimal point with hledger
- no country in the world uses space as decimal point
- we fairly recently added the ability to use space (just space, not all whitespace, hopefully) as digit group (eg thousands) separator
- consequently our existing logic allows it to be tried it as a decimal point as well
- we should change the logic to never try space as decimal point

**Comment by awjchen** [comment]
Agreed. I can make this change.
